### ISSUE TITLE
missing lzma support in borg-macosx64 binary

### ISSUE THREAD

**Issue by aravindhsampath** [borgbackup/borg#933]
borg-macosx64 1.0.0 binary

I'm trying to create a repo with lzma compression as per the docs. 

`borg init --encryption=keyfile ~/borg_backup/europe_photos`

`borg create -s -p --exclude-caches -C lzma ~/borg_backup/europe_photos/::17-apr-2015 ~/Pictures/Europe_trip_pass#1/`
I get: 
```
Enter passphrase for key /Users/asampath/.config/borg/keys/Users_asampath_borg_backup_europe_photos:
Local Exception.
Traceback (most recent call last):
  File "borg/archiver.py", line 1466, in main
  File "borg/archiver.py", line 1420, in run
  File "borg/archiver.py", line 211, in do_create
  File "borg/compress.pyx", line 183, in borg.compress.Compressor.__init__ (borg/compress.c:3806)
  File "borg/compress.pyx", line 173, in borg.compress.get_compressor (borg/compress.c:3651)
  File "borg/compress.pyx", line 122, in borg.compress.LZMA.__init__ (borg/compress.c:2435)
ValueError: No lzma support found.

Platform: Darwin Aravindhs-MacBook-Pro.local 15.4.0 Darwin Kernel Version 15.4.0: Fri Feb 26 22:08:05 PST 2016; root:xnu-3248.40.184~3/RELEASE_X86_64 x86_64 i386
Borg: 1.0.0  Python: Python 3.5.1
PID: 24599  CWD: /Users/asampath
sys.argv: ['borg', 'create', '-s', '-p', '--exclude-caches', '-C', 'lzma', '/Users/asampath/borg_backup/europe_photos/::17-apr-2015', '/Users/asampath/Pictures/Europe_trip_pass#1/']
SSH_ORIGINAL_COMMAND: None
```
I do have lzma installed and available.

```
Aravindhs-MacBook-Pro:~ asampath$ lama -V
xz (XZ Utils) 5.2.2
liblzma 5.2.2
Aravindhs-MacBook-Pro:~ asampath$ pip show pylzma
---
Metadata-Version: 2.0
Name: pylzma
Version: 0.4.8
Summary: Python bindings for the LZMA library by Igor Pavlov.
Home-page: http://www.joachim-bauch.de/projects/pylzma/
Author: Joachim Bauch
Author-email: mail@joachim-bauch.de
Installer: pip
License: LGPL
Location: /usr/local/lib/python2.7/site-packages
Requires:
Classifiers:
  Development Status :: 5 - Production/Stable
  Programming Language :: Python
  Programming Language :: Python :: 3
  Topic :: Software Development :: Libraries :: Python Modules
  Intended Audience :: Developers
  License :: OSI Approved :: GNU Library or Lesser General Public License (LGPL)
  Operating System :: OS Independent
```

Can anybody help me figure this out?

**Comment by enkore** [comment]
> Location: /usr/local/lib/**python2.7**/site-packages

LZMA is also part of the Python standard library. Maybe your Python 3.5.1 was compiled without LZMA?

**Comment by ThomasWaldmann** [comment]
you don't need pylzma bindings package btw.

**Comment by aravindhsampath** [comment]
@enkore ,
Thanks for pointing it out. I'm playing around with borg before choosing it as my backup solution for an set of machines. The directory with Pictures is just test data that I knew had a bunch of duplicate files. On my final setup, I'll plan to skip the compression option for my photos directory to speed things up. Thanks! 

Oh, that was dumb of me to check pip info when my default installation of Python is Python 2.7.x and Borg uses the installed python3.5.1

I checked for lzma on the Python installation that borg is using: 

```
Aravindhs-MacBook-Pro:~ asampath$ python3
Python 3.5.1 (default, Jan 22 2016, 08:54:32)
[GCC 4.2.1 Compatible Apple LLVM 7.0.2 (clang-700.1.81)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import lzma
>>> dir(lzma)
['CHECK_CRC32', 'CHECK_CRC64', 'CHECK_ID_MAX', 'CHECK_NONE', 'CHECK_SHA256', 'CHECK_UNKNOWN', 'FILTER_ARM', 'FILTER_ARMTHUMB', 'FILTER_DELTA', 'FILTER_IA64', 'FILTER_LZMA1', 'FILTER_LZMA2', 'FILTER_POWERPC', 'FILTER_SPARC', 'FILTER_X86', 'FORMAT_ALONE', 'FORMAT_AUTO', 'FORMAT_RAW', 'FORMAT_XZ', 'LZMACompressor', 'LZMADecompressor', 'LZMAError', 'LZMAFile', 'MF_BT2', 'MF_BT3', 'MF_BT4', 'MF_HC3', 'MF_HC4', 'MODE_FAST', 'MODE_NORMAL', 'PRESET_DEFAULT', 'PRESET_EXTREME', '_MODE_CLOSED', '_MODE_READ', '_MODE_WRITE', '__all__', '__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__spec__', '_compression', '_decode_filter_properties', '_encode_filter_properties', 'builtins', 'compress', 'decompress', 'io', 'is_check_supported', 'open']
```

Compress.pyx seems to be checking at 
import lzma
and throw ImportError exception if it fails. But, I can import it manually as seen above.. 

Any more ideas for me to check? 
Thanks!

**Comment by ThomasWaldmann** [comment]
compress.pyx catches the ImportError and sets lzma = None.
If you try to use lzma in such a case, it will say "no lzma support found".

Can't explain what's going wrong here, it should work like this if you can import lzma interactively.
Did you build / install borg using precisely that python 3.5.1 interpreter?

**Comment by ThomasWaldmann** [comment]
BTW, there is borg 1.0.2 now.

**Comment by ThomasWaldmann** [comment]
A workaround is of course to not use lzma.

You do not want lzma anyway except if you backup to a remote backup server over a very slow connection or you have another reason to pay a high cpu time price for a little more compression.

If you have a local backup storage, lz4 is super fast. gzip compresses better, but is also slower. Still, a low-level gzip compression is still a lot faster than lzma.

**Comment by aravindhsampath** [comment]
I switched to using zlib after lzma threw an error. I wanted to post the trace incase it was a bug that needed to be fixed. I chose lzma just because the docs claimed it's suited to get highest compression for slow storage. 
 
I had Python 3.5.1 installed in the past using brew install python3
and installed Borg using brew cask install borgbackup
(I had my brew cask updated before install) perhaps brew binaries need to be updated to 1.0.2?

I'll try to compile borg 1.0.2 myself then.

Thanks for the response. Really appreciate the work that has gone into borg.

**Comment by ThomasWaldmann** [comment]
can you point me to the brew package you used? or the "formula"?

**Comment by aravindhsampath** [comment]
Here is the info from Brew:

```
Aravindhs-MacBook-Pro:~ asampath$ brew cask info borgbackup
borgbackup: 1.0.0
BorgBackup
https://borgbackup.readthedocs.org/en/stable/
Not installed
https://github.com/caskroom/homebrew-cask/blob/master/Casks/borgbackup.rb
==> Contents
  borg-macosx64 (binary)
```

**Comment by ThomasWaldmann** [comment]
ok, that explains 2 things:
- likely, the mac os x vagrant machine i used for building the pyinstaller standalone borg binary had a python 3.5.1 without lzma support for some reason - my fault, sorry.
- there is no newer brew package for 1.0.1 or 1.0.2 because of #913

**Comment by aravindhsampath** [comment]
Glad you were able to figure it out. 

I use OS X as my daily driver and will be happy to test new releases and build them. I may need some help setting up to do it the right way though. Would you consider that?

**Comment by ThomasWaldmann** [comment]
@aravindhsampath let's take this over to #913.

**Comment by ThomasWaldmann** [comment]
@aravindhsampath I updated github releases with borg-macosx64 1.0.2 - lzma works now, and also osxfuse was updated.

**Comment by ThomasWaldmann** [comment]
fixed in 1.0-maint, will be merged into master later.
